src/Pate/Verification/StrongestPosts.hs

```diff
@@ -223,38 +271,63 @@ visitNode (ReturnNode fPair) vars d gr0 =
    -- using the ordinary widening machinery.
 
    processReturn gr ret = withPair ret $ do
+     let vars = PS.scopeVars scope
      validState <- PVV.validInitState (Just ret) (PS.simVarState (PPa.pOriginal vars)) (PS.simVarState (PPa.pPatched vars))
      withAssumptionSet validState $
-       do bundle <- returnSiteBundle vars d ret
-          withPredomain bundle d $ do
+       do (asm, bundle) <- returnSiteBundle vars d ret
+          withAssumptionSet asm $ withPredomain bundle d $ do
             traceBundle bundle "Processing return edge"
     --        traceBundle bundle "== bundle asm =="
     --        traceBundle bundle (show (W4.printSymExpr asm))
-            widenAlongEdge bundle (ReturnNode fPair) d gr (GraphNode ret)
+            widenAlongEdge scope bundle (ReturnNode fPair) d gr (GraphNode ret)
 
 
 -- | Construct a "dummy" simulation bundle that basically just
 --   immediately returns the prestate as the poststate.
 --   This is used to compute widenings that propagate abstract domain
 --   information from function return nodes to the actual return sites.
+--   The only state update that occurs is that a fresh variable is created
+--   to represent the stack base in the output state.
+--   The returned 'PS.AssumptionSet' associates the internal stack base
+--   from the callee (i.e. in 'SimInput') with the stack
+--   pointer of the caller (i.e. in 'SimOutput').
+--   This effectively re-phrases the resulting domain from the
+--   function call with respect to the specific call site that we are
+--   returning, and allows us to properly contextualize any resulting inequalities
+--   with respect to what is visible in our stack frame.
```

Comment:
Is this making a strong non-interference assumption about the callee? Or is that what is explicitly being propagated to the caller here?

Reply:
No this isn't really making any assumptions about the callee, since any changes it has made (that result in inequalities in the domain) will appear in the incoming domain (i.e. the domain passed to 'visitNode').

Basically this is a "virtual" transition where all we are doing is taking the outgoing domain from the callee and re-phrasing it to be based off of the stack base of the caller. So yes, any changes will still be observable to the caller (assuming 'abstractOverVars' was successful during this widening step).